File: C#/Sudoku.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Coord = System.Int32;
using Block = System.Int32;
using Count = System.Int32;

class Sudoku
{
    static void Main()
    {
        string input;
        while ((input = Console.ReadLine()) != null)
            Console.WriteLine(Solve(input));
    }

    static string Solve(string input)
    {
        try
        {
            return new Sudoku().read(input).search().show();
        }
        catch (NoSolutionException)
        {
            return "NO SOLUTION";
        }
    }

    class NoSolutionException : Exception {}

    enum View { GRD, ROW, COL, BOX }
    enum State { OPEN, FIXED, BANNED }

    const int UNIT = 3;
    const int SIZE = UNIT * UNIT;
    const int COORD = SIZE * SIZE * SIZE;
    const int BLOCK = 4 * SIZE * SIZE;
    const Count DONE = SIZE + 1;
    const string DIGITS = "123456789";

    static Coord coord(int i, int j, int k) =>
        (i * SIZE + j) * SIZE + k;

    static Block block(View v, int p, int q) =>
        ((int)v * SIZE + p) * SIZE + q;

    static List<Block>[] parents = new List<Block>[COORD];
    static List<Coord>[] children = new List<Coord>[BLOCK];

    static Sudoku() {
        for (int i = 0; i < SIZE; i++)
            for (int j = 0; j < SIZE; j++)
            {
                int p = i / UNIT * UNIT + j / UNIT;
                for (int k = 0; k < SIZE; k++)
                {
                    var list = new List<Block>(4);
                    list.Add(block(View.GRD, i, j));
                    list.Add(block(View.ROW, i, k));
                    list.Add(block(View.COL, j, k));
                    list.Add(block(View.BOX, p, k));
                    parents[coord(i, j, k)] = list;
                }
            }
        for (Block b = 0; b < BLOCK; b++)
            children[b] = new List<Coord>(SIZE);
        for (Coord c = 0; c < COORD; c++)
            foreach (Block b in parents[c])
                children[b].Add(c);
    }

    State[] state = new State[COORD];
    Count[] count = new Count[BLOCK];
    Queue<Coord> queue = new Queue<Coord>();

    Sudoku()
    {
        for (Coord c = 0; c < COORD; c++)
            state[c] = State.OPEN;
        for (Block b = 0; b < BLOCK; b++)
            count[b] = SIZE;
    }

    Sudoku(Sudoku that)
    {
        for (Coord c = 0; c < COORD; c++)
            state[c] = that.state[c];
        for (Block b = 0; b < BLOCK; b++)
            count[b] = that.count[b];
    }

    Sudoku enqueue(Coord c)
    {
        queue.Enqueue(c);
        return this;
    }

    Sudoku read(string input)
    {
        int len = input.Length;
        for (int i = 0; i < SIZE; i++)
            for (int j = 0; j < SIZE; j++)
                if (i * SIZE + j < len)
                {
                    int k = DIGITS.IndexOf(input[i * SIZE + j]);
                    if (k != -1)
                        enqueue(coord(i, j, k));
                }
        return this;
    }

    string show()
    {
        char[] result = new char[SIZE * SIZE];
        for (int i = 0; i < SIZE; i++)
            for (int j = 0; j < SIZE; j++)
            {
                result[i * SIZE + j] = '.';
                for (int k = 0; k < SIZE; k++)
                    if (state[coord(i, j, k)] == State.FIXED)
                        result[i * SIZE + j] = DIGITS[k];
            }
        return new string(result);
    }

    Sudoku search()
    {
        while (queue.Count > 0)
        {
            Coord c = queue.Dequeue();
            if (state[c] == State.FIXED)
                continue;
            if (state[c] == State.BANNED)
                throw new NoSolutionException();
            fix(c);
        }

        Count min = count.Min();
        if (min == DONE)
            return this;

        Block b = Array.IndexOf(count, min);
        foreach (Coord c in children[b])
            if (state[c] == State.OPEN)
            {
                try
                {
                    return new Sudoku(this).enqueue(c).search();
                }
                catch (NoSolutionException)
                {}
            }

        throw new NoSolutionException();
    }

    void fix(Coord c)
    {
        state[c] = State.FIXED;
        foreach (Block b in parents[c])
            mark_done(b);
    }

    void mark_done(Block b)
    {
        count[b] = DONE;
        foreach (Coord c in children[b])
            if (state[c] == State.OPEN)
                ban(c);
    }

    void ban(Coord c)
    {
        state[c] = State.BANNED;
        foreach (Block b in parents[c])
            if (count[b] != DONE)
                countdown(b);
    }

    void countdown(Block b)
    {
        count[b]--;
        if (count[b] == 0)
            throw new NoSolutionException();
        if (count[b] == 1)
            foreach (Coord c in children[b])
                if (state[c] == State.OPEN)
                    enqueue(c);
    }
}

File: Java/Sudoku.java
import java.io.*;
import java.util.*;

class Sudoku {

    public static void main(String[] args) {
        try {
            BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in));
            String input;
            while ((input = in.readLine()) != null) {
                System.out.println(solve(input));
            }
        }
        catch (Exception e) {
            System.out.println(e.getMessage());
            System.exit(1);
        }
    }

    static String solve(String input) {
        try {
            return new Sudoku().read(input).search().show();
        }
        catch (NoSolutionException e) {
            return "NO SOLUTION";
        }
    }

    static class NoSolutionException extends Exception {
        private static final long serialVersionUID = 0;
    }

    static enum View { GRD, ROW, COL, BOX }
    static enum State { OPEN, FIXED, BANNED }

    static final int VIEW  = 4;
    static final int UNIT  = 3;
    static final int SIZE  = UNIT * UNIT;
    static final int COORD = SIZE * SIZE * SIZE;
    static final int BLOCK = VIEW * SIZE * SIZE;
    static final int DONE  = SIZE + 1;

    static final String DIGITS = "123456789";

    static List<List<Integer>> parents;
    static List<List<Integer>> children;

    static int coord(int i, int j, int k) {
        return (i * SIZE + j) * SIZE + k;
    }

    static int block(View v, int p, int q) {
        return (v.ordinal() * SIZE + p) * SIZE + q;
    }

    static {
        parents = new ArrayList<>(COORD);
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                int p = i / UNIT * UNIT + j / UNIT;
                for (int k = 0; k < SIZE; k++) {
                    parents.add(coord(i, j, k), Arrays.asList(
                                block(View.GRD, i, j),
                                block(View.ROW, i, k),
                                block(View.COL, j, k),
                                block(View.BOX, p, k)));
                }
            }
        }
        children = new ArrayList<>(BLOCK);
        for (int b = 0; b < BLOCK; b++) {
            children.add(b, new ArrayList<>(SIZE));
        }
        for (int c = 0; c < COORD; c++) {
            for (int b : parents.get(c)) {
                children.get(b).add(c);
            }
        }
    }

    State[]        state = new State[COORD];
    int[]          count = new int[BLOCK];
    Queue<Integer> queue = new LinkedList<>();

    Sudoku() {
        Arrays.fill(state, State.OPEN);
        Arrays.fill(count, SIZE);
    }

    Sudoku(Sudoku that) {
        System.arraycopy(that.state, 0, state, 0, COORD);
        System.arraycopy(that.count, 0, count, 0, BLOCK);
    }

    Sudoku enqueue(int c) {
        queue.add(c);
        return this;
    }

    Sudoku read(String input) {
        int len = input.length();
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                if (i * SIZE + j < len) {
                    int k = DIGITS.indexOf(input.charAt(i * SIZE + j));
                    if (k != -1) {
                        enqueue(coord(i, j, k));
                    }
                }
            }
        }
        return this;
    }

    String show() {
        char[] result = new char[SIZE * SIZE];
        Arrays.fill(result, '.');
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                for (int k = 0; k < SIZE; k++) {
                    if (state[coord(i, j, k)] == State.FIXED) {
                        result[i * SIZE + j] = DIGITS.charAt(k);
                    }
                }
            }
        }
        return new String(result);
    }

    Sudoku search() throws NoSolutionException {

        while (!queue.isEmpty()) {
            int c = queue.remove();
            switch (state[c]) {
                case OPEN:
                    fix(c);
                    break;
                case FIXED:
                    break;
                case BANNED:
                    throw new NoSolutionException();
            }
        }

        int min = DONE;
        int bmin = -1;
        for (int b = 0; b < BLOCK; b++) {
            if (count[b] < min) {
                min = count[b];
                bmin = b;
            }
        }

        if (min == DONE) {
            return this;
        }

        for (int c : children.get(bmin)) {
            if (state[c] == State.OPEN) {
                try {
                    return new Sudoku(this).enqueue(c).search();
                }
                catch (NoSolutionException e) {
                }
            }
        }

        throw new NoSolutionException();
    }

    void fix(int c) throws NoSolutionException {
        state[c] = State.FIXED;
        for (int b : parents.get(c)) {
            mark(b);
        }
    }

    void mark(int b) throws NoSolutionException {
        count[b] = DONE;
        for (int c : children.get(b)) {
            if (state[c] == State.OPEN) {
                ban(c);
            }
        }
    }

    void ban(int c) throws NoSolutionException {
        state[c] = State.BANNED;
        for (int b : parents.get(c)) {
            if (count[b] != DONE) {
                countdown(b);
            }
        }
    }

    void countdown(int b) throws NoSolutionException {
        count[b]--;
        if (count[b] == 0) {
            throw new NoSolutionException();
        }
        if (count[b] == 1) {
            for (int c : children.get(b)) {
                if (state[c] == State.OPEN) {
                    enqueue(c);
                }
            }
        }
    }
}
